ksql-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java

```diff
@@ -338,14 +333,9 @@ private KafkaStreams buildStreams(
         StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG,
         ksqlConfig.get(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG)
     );
-    if (ksqlConfig.get(KsqlConfig.KSQL_TIMESTAMP_COLUMN_INDEX) != null) {
```

Comment:
Yes i believe this is what we want. Each stream can specify the field it wants to use as the timestamp. The field that is used for the timestamp MUST be available on the input source so you can't do as you suggest in the final statement. That will not work as the column T2 doesn't exist on the input record.

Reply:
@dguy I agree with @rodesai . The `WITH` clause in the CSAS/CTAS describes the properties of the new stream or table that is being created and the source stream and table properties should not change.

Reply:
@hjafarpour the source is not changed. It is the new CSAS/CTAS that decides to use another timestampe

Reply:
The `timestamp` in `WITH` clause in CSAS/CTAS will set the sink timestamp so timestamp column that is set in this `WITH` should exist in the sink schema.
The timestamp for source is defined in the corresponding DDL statement's WITH clause. 
If a CSAS/CTAS statement need to use a source with a different timestamp than the one the source was created with we can do the following:
1. Either have a new DDL statement for the source first and set the timestamp column to the desired one.
2. Or have a intermediate CSAS/CTAS statement to create the source with the desired timestamp.
3. Or change the grammar and introduce `WITH` clause for `FROM`. This would be more complex if we have join.

Reply:
Unfortunately that is not how it works. The timestamp needs to exist in the source schema.
If you have the following:
`create stream bar (id bigint, c1 varchar, c2 bigint) with(kafka_topic='bar')`
and then you did
`create stream foo as select id c1, c2 + 1000 as timestamp  from bar with(timestamp=`timestamp`)`

The timestamp written to the topic `foo` would be the one from the record as consumed from the topic `bar`. This is because the topic being consume from, i.e., bar defines the schema, and the timestamp extractor that runs before the "select ..." happens, i.e., it is the first thing that runs post deserialization only sees the record as defined by `bar`, so the field timestamp doesn't exist.

Reply:
Today KSQL would extract whatever timestamp was specified in the WITH clause of the original ddl statement (for `bar`) (in this case just using the time from the metadata). If the user issued another query (e.g. `create stream baz as select * from foo`) then that would use the column `timestamp` from `foo`.

With the change in this PR, what happens if the timestamp column doesn't exist in the sink, and the sink is used in a subsequent query?

Reply:
In this case it wouldn't matter if the timestamp column doesn't exist in the sink. The timestamp is extracted from the input and then used as the timestamp that is written to the sink with the ProducerRecord.

Anyway, i'll change it back so that the timestamp is inherited from the previous create statement - though i'm not sure that is correct as if i create a stream with a timestamp i expect it to use that timestamp and not the one from the parent stream.

Reply:
Hey, I'm a little confused here.

What Damian has outlined makes the most sense to me. 

Given:
```
create stream foo (id bigint, timestamp bigint, blah varchar) ...
create stream bar WITH(timestamp='timestamp') as select * from foo;
create stream foo_bar as select * from foo;
```

It sounds like what @rodesai and @hjafarpour are saying is that the `WITH(timestamp)` clause on `bar` is not controlling the windowing of `bar`, but `foo_bar`. `bar` will have a ROWTIME of the raw record timestamp, and `foo_bar` will have a ROTIME defined by the `timestamp` value in the `bar` stream.

This feels really counter intuitive to me! 

With what I think @dguy is saying `bar` would have ROWTIME defined by the `timestamp` field in the value of the messages received. Which is how I would read this.

Also, what if I want to derive two streams from `foo` with different time stamp policies? With @dguy's approach I just have to add different `WITH(timestamp)` clauses:

```
create stream foo (id bigint, f1 bigint, f2 bigint) ...
create stream bar1 WITH(timestamp='f1') as select * from foo;
create stream bar2 WITH(timestamp='f2') as select * from foo;
```

Where as with the other approach I would what?  Need to define two different copies of `foo`? e.g. 

```
create stream foo1 (id bigint, f1 bigint, f2 bigint) ... WITH(timestamp=f1)...
create stream foo2 (id bigint, f1 bigint, f2 bigint) ... WITH(timestamp=f1)...
create stream bar1 as select * from foo1;
create stream bar2 as select * from foo2;
```

This would be very expensive if foo1 and foo2 where themselves CSAS operations. Doubling the query cost just to change timstamp policies of downstream streams seems unnecessary to me.

With Damian's approach, each statement is self contained. Where as with the latter approach, you need to find the statement of the source stream to understand the timestamp used on the derived stream. Changing the timestamp property of one stream doesn't affect that stream, but only derived streams... seems very strange to me, or am I misunderstanding?

Reply:
Here’s the way I understand it. I think the issue is that we are trying to take 2 different behaviors and combine them into one setting:
    1. The timestamp associated with a KSQL Stream or Table. A KSQL Stream or Table has a schema. It also has a TIMESTAMP setting that identifies which column in the schema is the timestamp for records in the Stream/Table. For any query that selects from a Stream or Table, the value in the column specified in the timestamp setting is what's used as the record timestamp
    2. The column from the source topic to use as a the record timestamp for a given query, overriding the timestamp setting for that KSQL Stream/Table.

Currently, the timestamp setting in CSAS/CTAS/CS/CT only does 1. I think this change had it do 2 as well for CSAS/CTAS. That might be ok, but then I have a few questions:
    - What happens if the timestamp specifies a column that doesn't exist in the Stream/Table being created by the CSAS/CTAS statement? Wouldn't any subsequent queries that select from that stream/table fail or not have timestamps because the timestamp column could not be found? If we want to collapse these 2 settings then we should validate the column exists in both the source and sink schema.
    - What should the semantics be if there is more than one source Stream/Table (e.g. in a join)?

> This would be very expensive if foo1 and foo2 where themselves CSAS operations. Doubling the query cost just to change timstamp policies of downstream streams seems unnecessary to me.

I agree that it would be more expensive. I think there may be better ways to solve the performance problem than merging the settings (e.g. have another setting for 2)

Reply:
>What happens if the timestamp specifies a column that doesn't exist in the Stream/Table being created by the CSAS/CTAS statement? Wouldn't any subsequent queries that select from that stream/table fail or not have timestamps because the timestamp column could not be found?

As long as the field existed on the source record it is fine. The timestamp would be extracted and set as the timestamp used by streams, hence will become the value of `ROWTIME`. The column doesn't need to exist in both source and sink.

> What should the semantics be if there is more than one source Stream/Table (e.g. in a join)?
In a join the timestamp is derived from the stream side as it the stream that triggers the output

Reply:
Hey @rodesai,

Yep, I can see the two operations here. I think, for the sake of this PR, that semantics should be left how they were. 

However, I think Damian's approach, while needing more thought, is still the more intuitive for users. For me, the `TIMESTAMP` option in a CSAS or CTAS statement should override the timestamp used in the query itself. 

Plus I'm struggling to think of a use-case where the current TIMESTAMP semantics within a CSAS would be useful. Can you? It seems that such a statement is saying 'I'm going to run the query using the timestamp provided by the source stream, but I want downstream queries to use a different timestamp'.   This seems a very strange thing to want to do.

I'll create another issue to track...

Reply:
I've created #836

Reply:
Being able to set the source timestamp instead of relying on the source declaration is a valid point and I agree with it but the problem is that the `WITH` clause in CTAS/CSAS statements is used to set not only timestamp but also other the properties of the result stream/table and including the properties for the source in the `WITH` clause would result in confusion. Also I think when we have stream/stream join as @rodesai said this would cause problem. Although in current Stream/Table join timestamp come from the stream side as @dguy said but stream/stream join is different (@dguy can you clarify this?) As I suggested above one way to address this would be to add `WITH` clause to the `FROM` clause so we can specify source properties too. Fo instance this would be an example CSAS statement with such WITH clause for source:
`CREATE STREAM S2 WITH ( kafka_topic='test') AS SELECT col1, col2 FROM sourceStream WITH (timestamp = 'timestamp_col');`

As you see we will have `WITH` clause for both source and sink. In case of join we can have `WITH` clause for each source too. This will require change in our current grammar.

Reply:
Yep, I think the second `WITH` clause is a good option. I think it'll still cause confusion initially, but is pretty intuitive once the difference is explained.  

It's just a shame `WHERE` is already taken, as it would work well instead of the second `WITH` :-/

Reply:
@hjafarpour - in KStreams the timestamp of a stream/stream join will come from the side that triggered the join.